allspice/utils/bom_generation.py

```diff
@@ -291,25 +233,28 @@ def _extract_repetitions(sheet_refs: list[dict]) -> dict[str, int]:
     return repetitions
 
 
-def _schdoc_component_from_attributes(
-    attributes: dict,
-    mapper: AttributesMapping,
-) -> SchematicComponent:
+def _component_attributes(component: dict) -> dict[str, str]:
     """
-    Make a SchDoc Component object out of the JSON for it in the SchDoc file.
+    Extract the attributes of a component into a dict.
 
-    :param attributes: The attributes for the component in the SchDoc
-    :param mapper: An AttributesMapping object, see the documentation for
-        AttributesMapping for more information.
-    :returns: A SchodocComponent object representing that component.
+    This also adds two properties of the component that are not attributes into
+    the dict.
     """
 
-    return SchematicComponent(
-        description=_find_first_matching_key(mapper.description, attributes),
-        designator=_find_first_matching_key(mapper.designator, attributes) or "",
-        manufacturer=_find_first_matching_key(mapper.manufacturer, attributes),
-        part_number=_find_first_matching_key(mapper.part_number, attributes),
-    )
+    attributes = {}
+
+    for key, value in component["attributes"].items():
+        attributes[key] = value["text"]
+
+    # The properties `part_id` and `description` are present in the top level of
+    # the component.
+    try:
+        attributes["__part_id"] = component["part_id"]
+        attributes["__description"] = component["description"]
```

Comment:
Is there any reason we shouldn't remove the underscores here? Essentially we're just flattening the attributes for components. 

Perhaps our data model is just wrong here, either in allspice tools or py-allspice. The reason we did this was essentially to separate required, standard "attributes" for components from variable, user-defined attribute keys. However, I can see the user wanting to access any required variables in much the same way. For example, maybe want components X,Y position. 

What if we let the user specify the path to the property they want to use?

Eg. `attributes.Value`, `part_id`, etc...

Reply:
the downside is that exposes our data-structure to the users, and means we need to stabilize it.

Reply:
The underscores make it so that it doesn't accidentally overwrite a user-set attribute. E.g. if a user sets an attribute called `description` this would overwrite that attribute.

I don't like the idea of letting users select paths because that would strongly tie users to our JSON format for components. For example, `part_id` isn't the actual term used in Altium, which AIUI is Library Reference. That is very risky business, especially as we're planning on switching from Tools to crysknife at least at some point in the future, and switch the format even within tools. We can version py-allspice, but we can't version people's `attributes_mapping.json`.

Either way, I don't think the path is simpler for people to use, and they'd have to learn what keys are available either way. This is even more complex as a generic path system would mean users have to type `attributes.Value.text` and not just `attributes.Value`, which is again confusing unless you read the component JSON to figure out what you can do. If we special case attributes such that `attributes.Value` would automatically extract the text, that would be even more confusing, and you would not be able to extract other semantic data from the attribute.

Reply:
In Altium the user cannot set an attribute named "description" because it's a static attribute (that's why we separate them in the IR), and that attribute exists for all components. However, I'm not sure what that looks like in other tools. I'm fine keeping the underscore, but that needs to be clearly stated in the documentation because it will be a point of confusion for users.

With versioned releases we will be able to make breaking changes when needed. Users can choose to update their py-allspice version or continue to use the previous version as needed. We can include upgrade guides if needed. Because we don't yet version the parser (crysknife) I'm fine with keeping the `__` prefix here and limiting the scope of attributes users can select (so that py-allspice is effectively defining the interface). Ultimately, we're going to need to make these utilities more generic, which will require us to expose the IR. 

When we do version-lock crysknife, we can have py-allspice call the appropriate crysknife version as well. 

This will take time to work with users & we'll have to adjust the IR to get it into a stable state, but the goal more broadly is to standardize around an interface that makes getting data from hardware designs consistent and flexible. The less we special-case interfaces for different utilities and CAD formats, the more consistent it will be as we build out this framework.

Reply:
Fully agree - once we've stabilized the JSON interface we can set it up with JSONPath and good docs so users can have more flexibility in how they can consume it. This will also (IMO) lead to better solutions to the architecture problem you were talking about, as we can have a clean client API in py-allspice to consuming the JSON without having to tie it to generating the BOM. This is also better for making it tool agnostic, because ideally the IR is tool agnostic!